### PR TITLE
Added pg pong

### PR DESCRIPTION
**Pull request by 3141592654**
@mhrmm Uploaded version of pg pong with comments and different variable names. It also no longer prints out results when it is told it should not display. This is because such a scenario probably occurs when it is training overnight and in such a scenario printing results to the screen would slow down the training. Also included the output for one iteration of training. This doesn't do very much, but it does demonstrate that it can be done.

### CODE REVIEW

File: pg-pong.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+""" Trains an agent with (stochastic) Policy Gradients on Pong. Uses OpenAI Gym.
+    From http://karpathy.github.io/2016/05/31/rl/. Additional comments added, variables renamed. """
+import numpy as np
+import cPickle as pickle
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    import pickle

File: pg-pong.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+# model initialization
+D = 80 * 80 # input dimensionality: 80x80 grid
+if resume:
+  model = pickle.load(open('save.p', 'rb'))
+else:
+  model = {}
+  model['W1'] = np.random.randn(H,D) / np.sqrt(D) # "Xavier" initialization
+  model['W2'] = np.random.randn(H) / np.sqrt(H)
+  
+grad_buffer = { k : np.zeros_like(v) for k,v in model.iteritems() } # update buffers that add up gradients over a batch
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    grad_buffer = { k : np.zeros_like(v) for k,v in model.items() }

File: pg-pong.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+rmsprop_cache = { k : np.zeros_like(v) for k,v in model.iteritems() } # rmsprop memory
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    rmsprop_cache = { k : np.zeros_like(v) for k,v in model.items() }

File: pg-pong.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                                              # Magic is what happens in Harry Potter. 
+    # The particular "gradient" chosen means that the changes in the weights are bigger if we pick the action that 
+    # was less likely, and also bigger if hidden layer outputs are bigger.
+
+    grad = policy_backward(hidden_layer_outputs, so_called_gradients)
+    for k in model: grad_buffer[k] += grad[k] # accumulate grad over batch
+
+    # perform rmsprop parameter update every batch_size episodes
+    if episode_number % batch_size == 0:
+      for k,v in model.iteritems():
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    for k,v in model.items():

File: pg-pong.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        g = grad_buffer[k] # gradient
+        rmsprop_cache[k] = decay_rate * rmsprop_cache[k] + (1 - decay_rate) * g**2
+        model[k] += learning_rate * g / (np.sqrt(rmsprop_cache[k]) + 1e-5)
+        grad_buffer[k] = np.zeros_like(v) # reset batch gradient buffer
+
+    # boring book-keeping
+    running_reward = reward_sum if running_reward is None else running_reward * 0.99 + reward_sum * 0.01
+    # I added the if render statement. If we are training the model, we don't want to waste time printing things out
+    if render: print 'resetting env. episode reward total was %f. running mean: %f' % (reward_sum, running_reward)
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    if render: print('resetting env. episode reward total was %f. running mean: %f' % (reward_sum, running_reward))

File: pg-pong.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    # Save the data every hundred observations. Why every hundred? Perhaps Andrej liked 100.
+    if episode_number % 100 == 0: pickle.dump(model, open('save.p', 'wb'))
+    reward_sum = 0
+    observation = env.reset() # reset env
+    prev_x = None
+  #I added this if render statement for the same reason
+  if render:
+    if reward != 0: # Pong has either +1 or -1 reward exactly when game ends.
+      print ('ep %d: game finished, reward: %f' % (episode_number, reward)) + ('' if reward == -1 else ' !!!!!!!!')
```
Comment (mhrmm) [inline pg-pong.py]: Python 3:

    print(('ep %d: game finished, reward: %f' % (episode_number, reward)) + ('' if reward == -1 else ' !!!!!!!!'))

**Review by mhrmm — COMMENTED** [review]
LGTM. Before merging, let's just update the code to be compatible with Python 3. I've indicated the necessary changes.
